Fixes Issue #277. FCSData objects now able to be (de)pickled.

SynbioLucas opened the pull request:
Added code as described in Issue #277 to enable FCSData objects to be pickled and depickled. Wrote a test to verify that this is working properly and that depickled objects contain the correct attributes. All tests pass with no errors.

SynbioLucas committed:
Fixed Issue #277. FCSData objects now pickle-able

Fixes Issue #277. FCSData objects can now be correctly serialized and
deserialized with Pickle.

Added:
- New unit test in test_io.py that correctly catches the issue and now
  runs cleanly.
- Added __reduce__ and __setstate__ methods to FCSData class to allow
  correct Pickle behavior, as documented in the Issue.

SynbioLucas committed:
Expanded test slightly

castillohair commented on FlowCal/io.py:
```diff
@@ -1760,6 +1760,38 @@ def __array_finalize__(self, obj):
             self._resolution = copy.deepcopy(obj._resolution)
 
+    # override __reduce__ and __setstate__ to fix Issue #277
+    # Adapted from https://stackoverflow.com/a/26599346
```
Remove these. I would rather have this explained in a docstring in each function with more detail. See below.

castillohair commented on FlowCal/io.py:
```diff
@@ -1760,6 +1760,38 @@ def __array_finalize__(self, obj):
 
+    # override __reduce__ and __setstate__ to fix Issue #277
+    # Adapted from https://stackoverflow.com/a/26599346
+    def __reduce__(self):
```
We need a docstring here explaining what this function does and why it was replaced. It doesn't have to be too detailed, as this function will not be exposed to the user, but it should be enough to be useful for anyone in the future modifying this code. Please refer to https://github.com/numpy/numpy/blob/master/doc/example.py.

JS3xton replied:
Consider naming the returned parameters to hint at what they are, e.g.:

        Returns
        -------
        init : callable
            Callable used to recreate the inital instance
        init_args : tuple
            Arguments to be passed to the callable to reconstitute the
            FCSData object
        state : tuple
            Internal state data needed to fully reconstruct the instance. Data
            is passed to self.__setstate__() to reconstruct object during
            deserialization. The order of elements must match that expected
            by __setstate__.

Also, fix the typo with "inital"

castillohair replied:
[Name of return values is optional](https://numpydoc.readthedocs.io/en/latest/format.html), as they're not exposed to the method caller anyway.

castillohair commented on FlowCal/io.py:
```diff
@@ -1760,6 +1760,38 @@ def __array_finalize__(self, obj):
+        # Return updated tuple with full set_state tuple
+        return (pickled_state[0], pickled_state[1], new_state)
+
+    def __setstate__(self, state):
```
We need a docstring here explaining what this function does and why it was replaced. It doesn't have to be too detailed, as this function will not be exposed to the user, but it should be enough to be useful for anyone in the future modifying this code. Please refer to https://github.com/numpy/numpy/blob/master/doc/example.py.

castillohair commented on test/test_io.py:
```diff
@@ -2681,6 +2682,44 @@ def test_mean_axis(self):
+            '_acquisition_end_time', '_acquisition_start_time', '_time_step'
+            '_data_type', '_analysis', '_text', '_infile']
+        for attr in pickle_sensitive_attrs:
+            if hasattr(self.d, attr):
```
Why is this necessary? You should know in advance if these attributes are valid or not.

SynbioLucas replied:
Yep, you're right - this is now deleted.

castillohair commented on test/test_io.py:
```diff
@@ -2681,6 +2682,44 @@ def test_mean_axis(self):
+        with open(self.test_file, 'r') as infile:
+            test_fcs = cPickle.load(infile) # de-serialization
+
+        # Check all attributes/methods are preserved
```
This section is checking that the same attributes exist, and that some subset of the hidden attributes have the same value. I would recommend that the contents of the array and the actual values of computed attributes are also checked. Even if you know that these worked before this PR, it would be a good idea for the tests to be as comprehensive as possible in order to detect future code changes that break proper pickling.

For the array values, you might want to use [`assert_array_equal`](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.testing.assert_array_equal.html) or [`assert_almost_equal`](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.testing.assert_almost_equal.html).

SynbioLucas replied:
These are good ideas - thanks! I believe I've incorporated your suggestions, but please let me know if there are still cases/attributes/data that I'm forgetting!

castillohair reviewed (changes requested):
Thanks for this PR. Indeed all old tests still pass and the Excel UI still works with the example file.

I'm requesting docstrings in the two functions that were added to `FCSData`, and a few additions to the test function. After that I think it should be ready to merge, unless @JS3xton has anything else to say.

SynbioLucas committed:
Updated FCSData reduce, setstate docs, test

- Added numpydoc-compliant docstrings to FCSData.__reduce__() and
.__setstate__().
- Added additional test cases to TestFCSDataPickle to include derived
attribute values, function returns, and self array data.

SynbioLucas committed:
Merge remote-tracking branch 'upstream/master' into pickle-fix

castillohair committed:
Merge branch 'develop' into pickle-fix

castillohair committed:
Updated to support python3.

JS3xton commented on test/test_io.py:
```diff
@@ -4,8 +4,13 @@
+try:
+   import cPickle as pickle
+except:
+   import pickle
```
This `except` statement should be as precise as possible. On my computer, you should except a `ModuleNotFoundError`.

JS3xton commented on FlowCal/io.py:
```diff
@@ -1760,6 +1760,97 @@ def __array_finalize__(self, obj):
+        # NDArray params.
+        new_state = pickled_state[2] + fcsdata_attrs
+
+        return (pickled_state[0], pickled_state[1], new_state)
```
These variable names were confusing to me. References to "state" appear most appropriate for the 3rd tuple element; calling the entire `__reduce__` tuple `pickled_state` therefore sounds imprecise to me.

The [Python 3 docs](https://docs.python.org/3.7/library/pickle.html#object.__reduce__) suggest the output of `__reduce__` is called a "reduce value":
> The __reduce__() method takes no argument and shall return either a string or preferably a tuple (the returned object is often referred to as the “reduce value”).

So consider changing `pickled_state` to something like `super_reduce_value`.

---

Also, the docs suggest that it's possible for `__reduce__` to return up to a 5-tuple, which this code would currently truncate. Can we guarantee that `np.ndarray` will never use those extra tuple elements? Or that it will always already have a "state" element in the 3rd position? I prefer a more precise implementation that is robust to `__reduce__` outputs with between 2 and 5 elements, as the docs allow for.

JS3xton commented on FlowCal/io.py:
```diff
@@ -1760,6 +1760,97 @@ def __array_finalize__(self, obj):
+            self._amplifier_gain, self._range, self._resolution,)
+        # Append these FCSData-specific state parameters to the existing
+        # NDArray params.
+        new_state = pickled_state[2] + fcsdata_attrs
```
The [docs](https://docs.python.org/3.7/library/pickle.html#object.__reduce__) don't specify how the "state" information is  stored (i.e. in general, we don't know what kind of object `pickled_state[2]` is). For now, the `+` operator appears to append our state information to the parent's state information, but I think it would be more robust to layer our own state information around the parent state information without making any assumptions about the parent state information. E.g. `new_state = (pickled_state[2],fcsdata_attrs)`. Our `__setstate__` function would therefore set our attributes from `new_state[1]` and then pass `new_state[0]` to the `super.__setstate__`.

If we're separating our own state information, you might also consider using a `namedtuple` or `dict` to avoid relying on positional indices to get the attributes unmixed correctly.

JS3xton commented on FlowCal/io.py:
```diff
@@ -1760,6 +1760,97 @@ def __array_finalize__(self, obj):
+        self._infile = state[-13]
+        # To set remaining state data for parent class, call its
+        # __setstate__ with remaining attrs
+        super(FCSData, self).__setstate__(state[0:-13])
```
As mentioned in `__reduce__`, consider a more robust implementation that packages our state information separate from the parent state information and uses better data structures to avoid relying on position/index to unpack our attributes.

JS3xton replied:
Also, does it matter who does `__setstate__` first? Practically, I think it doesn't matter for this case (?), but philosophically, I kind of feel like the parent class should go first, but I'm not sure.

JS3xton committed:
Clarified documentation regarding FCSData pickling.

Named return parameters of __reduce__ function.

JS3xton committed:
Made try/except statement more precise.

JS3xton committed:
Made FCSData pickling more robust.

-FCSData pickling makes fewer assumptions about the pickling data
 structures created by its parent class (numpy's NDArray).
-Added data structure to store FCSData state information
 (_FCSDataPickleState namedtuple), which circumvents relying on
 matching index positions to identify attributes during unpickling.
-Made minor modifications to the documentation.
